Source/Libraries/openXDA.APIAuthentication/APIQuery.cs

```diff
@@ -0,0 +1,260 @@
+        public APIQuery(string apiKey, string apiToken, IEnumerable<string> hostURLs)
+        {
+            APIKey = apiKey;
+            APIToken = apiToken;
+
+            Hosts = hostURLs
+                .Select(url => new Host(url))
+                .ToList();
+
+            // Select a random host for the first API call attempt
+            if (Hosts.Count > 1)
+                Interlocked.Exchange(ref m_hostIndex, InitialHostIndex);
+        }
+
+        #endregion
+
+        #region[ Properties ]
+
+        /// <summary>
+        /// The API key identifying the user of the API.
+        /// </summary>
+        public string APIKey { get; }
+
+        /// <summary>
+        /// The API token used to authenticate the user.
+        /// </summary>
+        public string APIToken { get; }
+
+        /// <summary>
+        /// The list of URLs that locate the hosts providing access to the API.
+        /// </summary>
+        public IEnumerable<string> HostsURLs => Hosts
+            .Select(host => host.URL);
+
+        private List<Host> Hosts { get; }
+
+        private int InitialHostIndex => Tuple
+            .Create(Environment.TickCount, Environment.MachineName)
+            .GetHashCode() % Hosts.Count;
+
+        #endregion
+
+        #region [ Methods ]
+
+        /// <summary>
+        /// Sends a web request to the host using the credentials for API authentication.
+        /// </summary>
+        /// <param name="configure">Action that configures the HTTP request.</param>
+        /// <param name="path">Path to the API endpoint locating the resource to be requested.</param>
+        /// <param name="cancellationToken">Token used to cancel the request before it has completed.</param>
+        /// <returns>The HTTP response returned by the host that handled the request.</returns>
+        public async Task<HttpResponseMessage> SendWebRequestAsync(Action<HttpRequestMessage> configure, string path, CancellationToken cancellationToken = default)
+        {
+            int initialHostIndex = Interlocked.CompareExchange(ref m_hostIndex, 0, 0);
+
+            void UpdateHostIndex(int hostIndex) =>
+                Interlocked.CompareExchange(ref m_hostIndex, hostIndex, initialHostIndex);
+
+            for (int i = 0; i < Hosts.Count; i++)
+            {
+                int hostIndex = (initialHostIndex + i) % Hosts.Count;
+                Host host = Hosts[hostIndex];
+
+                try
+                {
+                    HttpResponseMessage response = await SendWebRequestToAsync(host, configure, path, cancellationToken);
+                    UpdateHostIndex(hostIndex);
+                    return response;
+                }
+                catch (HostUnreachableException)
+                {
+                    continue;
+                }
+                catch
+                {
+                    UpdateHostIndex(hostIndex);
+                    throw;
```

Comment:
I suppose it might be a good idea to wrap this one so we can include the host URL in the error message. Otherwise, identifying the failed host could be a challenge.